controllers/notebook/notebook_controller.go

```diff
@@ -0,0 +1,506 @@
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/intstr"
+	"k8s.io/client-go/tools/record"
+	"k8s.io/kubernetes/pkg/apis/core"
+	"k8s.io/utils/pointer"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	notebookv1alpha1 "github.com/alibaba/kubedl/apis/notebook/v1alpha1"
+)
+
+func NewNotebookController(mgr ctrl.Manager, _ options.JobControllerConfiguration) *NotebookReconciler {
+	r := &NotebookReconciler{
+		Client: mgr.GetClient(),
+		Log:    ctrl.Log.WithName("controllers").WithName("Notebook"),
+		Scheme: mgr.GetScheme(),
+	}
+	r.recorder = mgr.GetEventRecorderFor("NotebookController")
+	return r
+}
+
+// NotebookReconciler reconciles a Notebook object
+type NotebookReconciler struct {
+	client.Client
+	Log      logr.Logger
+	Scheme   *runtime.Scheme
+	recorder record.EventRecorder
+}
+
+// +kubebuilder:rbac:groups=notebook.kubedl.io,resources=notebooks,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=notebook.kubedl.io,resources=notebooks/status,verbs=get;update;patch
+// +kubebuilder:rbac:groups="",resources=pods,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups="",resources=pods/status,verbs=get;update;patch
+// +kubebuilder:rbac:groups="",resources=services,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups="",resources=services/status,verbs=get;update;patch
+// +kubebuilder:rbac:groups="networking.k8s.io",resources=ingresses,verbs=get;list;watch;create;update;patch;delete
+
+func (r *NotebookReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
+	_ = r.Log.WithValues("notebook", req.NamespacedName)
+
+	notebook := &notebookv1alpha1.Notebook{}
+	err := r.Get(context.Background(), req.NamespacedName, notebook)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			r.Log.Info("notebook doesn't exist", "name", req.String())
+			return reconcile.Result{}, nil
+		}
+		r.Log.Error(err, "fail to get notebook", "name", req.String())
+		return reconcile.Result{}, err
+	}
+
+	if notebook.Status.Condition == notebookv1alpha1.NotebookTerminated {
+		r.Log.Info("notebook terminated", "name", req.String())
+		return reconcile.Result{}, nil
+	}
+
+	// 1. create the notebook pod if not existing
+	pod := &corev1.Pod{}
+	nbPodName := types.NamespacedName{Namespace: notebook.Namespace, Name: nbNamePrefix(notebook.Name)}
+	err = r.Get(context.Background(), nbPodName, pod)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			r.Log.Info("create notebook pod " + nbPodName.String())
+			pod.ObjectMeta = *notebook.Spec.Template.ObjectMeta.DeepCopy()
+			pod.Name = nbPodName.Name
+			pod.Namespace = nbPodName.Namespace
+			if pod.Labels == nil {
+				pod.Labels = make(map[string]string, 1)
+			}
+			pod.Labels[notebookv1alpha1.NotebookNameLabel] = notebook.Name
+			pod.Spec = notebook.Spec.Template.Spec
+
+			addNotebookDefaultPort(&pod.Spec)
```

Comment:
this and other defaulter functions can be moved to `_default.go` and invoked by `scheme.Default()`?